controllers/operator_controller.go

```diff
@@ -47,7 +51,48 @@ type OperatorReconciler struct {
 // For more details, check Reconcile and its Result here:
 // - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.13.1/pkg/reconcile
 func (r *OperatorReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
-	_ = log.FromContext(ctx)
+	l := log.FromContext(ctx).WithName("reconcile")
+	l.V(1).Info("starting")
+	defer l.V(1).Info("ending")
+
+	var existingOp = &operatorsv1alpha1.Operator{}
+	if err := r.Get(ctx, req.NamespacedName, existingOp); err != nil {
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	reconciledOp := existingOp.DeepCopy()
+	res, reconcileErr := r.reconcile(ctx, reconciledOp)
+
+	// Compare status and update if required
+	if !equality.Semantic.DeepEqual(existingOp.Status, reconciledOp.Status) {
+		if updateErr := r.Status().Update(ctx, reconciledOp); updateErr != nil {
+			return res, utilerrors.NewAggregate([]error{reconcileErr, updateErr})
+		}
+	}
+
+	// Compare resources - ignoring status - and update if required
+	existingOp.Status, reconciledOp.Status = operatorsv1alpha1.OperatorStatus{}, operatorsv1alpha1.OperatorStatus{}
+	if !equality.Semantic.DeepEqual(existingOp, reconciledOp) {
+		if updateErr := r.Update(ctx, reconciledOp); updateErr != nil {
+			return res, utilerrors.NewAggregate([]error{reconcileErr, updateErr})
+		}
+	}
```

Comment:
Could you explain when the controller would modify something other than the Operator CR's status? I can't imagine a scenario where we'd want to modify the resource's spec, is this to update something in the metadata (annotations, labels, etc..)? If the point of this update call is to modify something outside the spec, does it make sense to exclude it?

Reply:
This was suggested by @joelanford; it's copied from the rukpak reconciler.

Reply:
(And possibly metadata)

Reply:
Finalizers are the only thing that comes to mind. Defaulting should be handled for us by the apiserver during admission. Not reconciliation.

This is a good question though. Perhaps we should have the code panic if it detects a change we made to non-status, non-finalizers, as that would indicate a programming mistake.

Reply:
Should rukpak be updated as well?

Reply:
@joelanford: something like?
```go
	existingOp.Status, reconciledOp.Status = operatorsv1alpha1.OperatorStatus{}, operatorsv1alpha1.OperatorStatus{}
	existingOp.Finalizers, reconciledOp.Finalizers = []string{}, []string{}
	if !equality.Semantic.DeepEqual(existingOp, reconciledOp) {
		panic("spec or metadata changed by reconciler")
	}

Reply:
This would strip finalizers from the object, which we may not want to. Something like this may work.
```
existingOp.Status, reconciledOp.Status = operatorsv1alpha1.OperatorStatus{}, operatorsv1alpha1.OperatorStatus{}
	if !equality.Semantic.DeepEqual(existingOp.Spec, reconciledOp.Spec) {
		panic("spec or metadata changed by reconciler")
	}
``

Reply:
Are you sure? We aren't doing an `r.Update()` in this case, although that might still be necessary due to updated finalizers. We can't just compare the `Specs`, as that ignores other `metadata`.

Reply:
OK, weirdness! `r.Status().Update()` actually makes additional modifications to the structure. So, do all the comparisons _first_, then update the status, `panic()` if required, then update the spec (should just be the finalizers).
The `reconciledOp` `existingOp` and `compareOp` are all local variables, and k8s is unmodified until an `Update()` call is made.

Reply:
>Are you sure? We aren't doing an r.Update() in this case, although that might still be necessary due to updated finalizers. We can't just compare the Specs, as that ignores other metadata.

Ah sorry, I didn't notice that you changed the code between the time that I initially reviewed the PR and made that comment.